src/hardware/system_timer.rs

```diff
@@ -103,15 +116,18 @@ impl Monotonic for SystemTimer {
         regs.cnt.reset();
     }
 
-    fn set_compare(&mut self, instant: &Instant<Self>) {
+    fn now(&mut self) -> Self::Instant {
+        Self::Instant::from_ticks(self.get_ticks())
+    }
+
+    fn set_compare(&mut self, instant: Self::Instant) {
         let regs = unsafe { &*hal::device::TIM15::ptr() };
 
-        let now = self.try_now().unwrap();
-        match now.checked_duration_until(instant) {
+        match instant.checked_duration_since(self.now()) {
```

Comment:
Instead of that entire match I would just:

```rust
let ticks = instant
    .checked_duration_since(self.now())
    .map_or(0, |t| t.ticks())
    .max(1)
    .min(0xffff);
regs.ccr1.write(|w| w.ccr().bits(ticks as _));
regs.dier.modify(|_, w| w.cc1ie().set_bit());
```

Reply:
That match is actually not equivalent to what's written here. We load the CCR1 (compare-capture-reg 1) with the u16 compare timestamp - the `map_or` function maps the Option to the tick delta between `instant` and `now`, which is not equivalent to the timestamp that we want to wake up at, but rather the time difference between the two instants.

We only use the delta-tick calculated value to determine whether or not the `instant` will occur within a single overflow of the timer. If it does, we just truncate the `instant` to a u16 tick count (since we guarantee that the 16 MSBs are irrelevant via the earlier check) and store that in CCR1. If it doesn't, we store u16::MAX in CCR1 to re-evaluate at the next timer overflow.

I think the correct chain would be:
```rust
let ticks = if instant.checked_duration_since(self.now()).unwrap_or(1) < (1 << 16) {
    instant.ticks() as u16
} else {
    u16::MAX
}
```

Reply:
However, after staring at the code for a while, I think this can be simplified further. Give me a moment.

Reply:
In any case, given that we plan on replacing this code soon with `systick-monotonic`, I'd recommend we leave it as-is.

Reply:
Ah. Yes. You have TIM15 free running and interrupt on match, not interrupt on wrap and reload.